File: DSA_2215/sort/merge/merge_sort.go
package main

import "fmt"

// // Example use
// unsortedList := []float64{38, 27, 43, 3, 9, 82, 10}

// fmt.Println("Unsorted List:", unsortedList)

// ascendingSorted := mergeSort(unsortedList, Ascending)
// descendingSorted := mergeSort(unsortedList, Descending)

// fmt.Println("Ascending Sorted List:", ascendingSorted)
// fmt.Println("Descending Sorted List:", descendingSorted)

type SortOrder int

const (
	Ascending SortOrder = iota
	Descending
)

func merge(left, right []float64, order SortOrder) []float64 {
	result := make([]float64, 0, len(left)+len(right))
	l, r := 0, 0

	for l < len(left) && r < len(right) {
		if (order == Ascending && left[l] < right[r]) ||
			(order == Descending && left[l] > right[r]) {
			result = append(result, left[l])
			l++
		} else {
			result = append(result, right[r])
			r++
		}
	}

	result = append(result, left[l:]...)
	result = append(result, right[r:]...)

	return result

}

func mergeSort(arr []float64, order SortOrder) []float64 {
	if len(arr) <= 1 {
		return arr
	}

	mid := len(arr) / 2
	left := mergeSort(arr[:mid], order)
	right := mergeSort(arr[mid:], order)

	return merge(left, right, order)
}

func main() {
	unsortedList := []float64{38, 27, 43, 3, 9, 82, 10}

	fmt.Println("Unsorted List:", unsortedList)

	ascendingSorted := mergeSort(unsortedList, Ascending)
	descendingSorted := mergeSort(unsortedList, Descending)

	fmt.Println("Ascending Sorted List:", ascendingSorted)
	fmt.Println("Descending Sorted List:", descendingSorted)
}

File: DSA_2215/sort/merge/merge_sort.dart
// Function to perform merge sort
List<int> mergeSort(List<int> list) {
  // Base case: If the list has one or zero elements, it's already sorted.
  if (list.length <= 1) {
    return list;
  }

  // Calculate the middle index of the list.
  final int mid = list.length ~/ 2;
  print("mid: $mid");

  // Recursively sort the left and right halves of the list.
  final List<int> left = mergeSort(list.sublist(0, mid));
  print("left: $left");
  final List<int> right = mergeSort(list.sublist(mid));
  print("right: $right");

  // Merge the sorted left and right halves.
  return merge(left, right);
}

// Function to merge two sorted lists into one sorted list.
List<int> merge(List<int> left, List<int> right) {
  final List<int> result = []; // The sorted result.

  int leftIndex = 0; // The index of the first unprocessed element in left.
  int rightIndex = 0; // The index of the first unprocessed element in right.

  // Compare elements from both lists and add the smaller element to the result.
  while (leftIndex < left.length && rightIndex < right.length) {
    if (left[leftIndex] < right[rightIndex]) {
      result.add(left[leftIndex]);
      leftIndex++;
    } else {
      result.add(right[rightIndex]);
      rightIndex++;
    }
  }

  // Add any remaining elements from the left and right lists to the result.
  result.addAll(left.sublist(leftIndex));
  result.addAll(right.sublist(rightIndex));

  return result;
}

void main() {
  // Example unsorted list to be sorted using merge sort.
  List<int> unsortedList = [38, 27, 43, 3, 9, 82, 10];

  // Call the mergeSort function to sort the list.
  List<int> sortedList = mergeSort(unsortedList);

  // Print the unsorted and sorted lists for demonstration.
  print("Unsorted List: $unsortedList");
  print("Sorted List: $sortedList");
}
